### ISSUE TITLE
Using both SearchOptions & Operator in Search query cause all indexed documents to be matched and returned

### ISSUE THREAD

**Issue by pmengal** [ravendb/ravendb#12750]
I use this 

`IList<Acticle> users = session
    .Query<Acticle>()
    .Search(x => x.Title, searchTerms, options: SearchOptions.Or, @operator: SearchOperator.And)
    .Search(x => x.Content, searchTerms, options: SearchOptions.Or, @operator: SearchOperator.And)
    .ToList();`

This will return ALL articles. When I remove one of the option (SearchOptions or Operator), the behavior is the one I expected.

I wanted to match article article that had the search terms in either the Title or Content, with the search terms having to be all matched. Example:

An article titled : "Using RavenDB to create a search engine" with content having the sentence "In order to use RavenDB as a search engine, you need to ...", would return no results if we used "RavenDB" or "Search" as the search term, and would match any document that contains both "RavenDB" and "Search" in either the Title or Content.

**Comment by ayende** [comment]
The `SearchOptions.Or` is actually the default, which means that what you are actually sending is something like this:

`from 'Articles' where search(Title, $p0) and search(Content, $p1)`

This query actually is saying that you need this to be in *both* fields, which I think isn't what you want.

You need to use `SearchOptions.And` and `SearchOperator.Or` (or discard it, that is the default).  I know that this is confusing, and I'm sorry about that. If you reverse the options / operator, you'll have the following query, instead:

`from 'Articles' where search(Title, $p0, and) or search(Content, $p1, and)`

That is probably easier to understand than when looking at the Linq API.

Note that you can always call `.ToString()` on the query to see what will be sent to the server, that can be very helpful for debugging.

**Comment by pmengal** [comment]
Thanks, I did all possible combinaisons with the ToString(). Actually, the issue with the interaction with the other where clause filters that I didn't copy in my original report. Because of the context of my query, the filter in question is filtering no records at all. So using the SearchOptions.Or was returning everything and I thought it was an issue with RavenDB.

For my use case, I needed to use SearchOperator.And without specifying SeachOptions as the default (Guess) is adding the extra parenthesis needed to isolate searches from other filters with the proper and operator. Here are the results of the tests:

For the tests $p3 is set to DateTime.MinValue & $p4 is set to DateTime.MaxValue.

SearchOperator default is Or
SearchOptions default is Guess

**SearchOptions.And & SearchOperator.Or**

`from index 'ArticleSearch' where search(Title, $p1) and search(Description, $p2) and PublicationDate > $p3 and PublicationDate < $p4`

Note what we needed...

**SearchOptions.And & SearchOperator.And**

`from index 'ArticleSearch' where search(Title, $p1, and) and search(Description, $p2, and) and (PublicationDate > $p3 and PublicationDate < $p4)`

Better, but not quite good...

**SearchOptions.Or & SearchOperator.And**

`from index 'ArticleSearch' where search(Title, $p1, and) or search(Description, $p2, and) or PublicationDate > $p3 and PublicationDate < $p4`

This will returns the entire index as other filters (that don't filter anything), are not isolated properly in the query.

**SearchOptions.Or & SearchOperator.Or**

`from index 'ArticleSearch' where search(Title, $p1) or search(Description, $p2) or PublicationDate > $p3 and PublicationDate < $p4`

In our context, this will also return everything.

**SearchOptions.Guess & SearchOperator.Or**

`from index 'ArticleSearch' where (search(Title, $p1) or search(Description, $p2)) and (PublicationDate > $p3 and PublicationDate < $p4)`

This is much better, but this will return too much records because each word in the search terms will trigger the match.

**SearchOptions.Guess & SearchOperator.And**

from index 'ArticleSearch' where (search(Title, $p1, and) or search(Description, $p2, and)) and (PublicationDate > $p3 and PublicationDate < $p4)

This is (almost) perfect!

**SearchOptions.And**

`from index 'ArticleSearch' where search(Title, $p1) and search(Description, $p2) and (PublicationDate > $p3 and PublicationDate < $p4)`

Same problems as above.

**SearchOptions.Or**

`from index 'ArticleSearch' where search(Title, $p1) or search(Description, $p2) or PublicationDate > $p3 and PublicationDate < $p4`

Same.

**SearchOptions.Guess**

`from index 'ArticleSearch' where (search(Title, $p1) or search(Description, $p2)) and (PublicationDate > $p3 and PublicationDate < $p4)`

Doesn't help.

**SearchOperator.Or**

`from index 'ArticleSearch' where (search(Title, $p1) or search(Description, $p2)) and (PublicationDate > $p3 and PublicationDate < $p4)`

Doesn't help either.

**SearchOperator.And**

`from index 'ArticleSearch' where (search(Title, $p1, and) or search(Description, $p2, and)) and (PublicationDate > $p3 and PublicationDate < $p4)`

This is perfect and what we needed. Specifying the SearchOperator.And (as the default is Or), and no SearchOptions as the desired value is the default (SearchOptions.Guess)
